petastorm/pytorch.py

```diff
@@ -367,3 +367,123 @@ def __enter__(self):
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self.reader = reader
+        self.batch_size = batch_size
+        self.transform_fn = transform_fn or torch.as_tensor
+        self.num_epochs = num_epochs
+        self.seed = seed
+        self.capacity = capacity
+        self.shuffle = shuffle
+        self.size = 0
+
+    def _load_rows_into_mem(self):
+        n_rows = 0
+        buffer_full = False
+        self.buffer = None
+        self.keys = None
+
+        for row in self.reader:
+            if buffer_full:
+                break
+            # Default collate does not work nicely on namedtuples and treat them as lists
+            # Using dict will result in the yielded structures being dicts as well
+            row_as_dict = row._asdict()
+
+            # Promote some types that are incompatible with pytorch to be pytorch friendly.
+            _sanitize_pytorch_types(row_as_dict)
+
+            for k, v in row_as_dict.items():
+                if not self.reader.batched_output:
+                    row_as_dict[k] = self.transform_fn([v])
+                else:
+                    row_as_dict[k] = self.transform_fn(v)
+
+            if not self.keys:
+                self.keys = row_as_dict.keys()
+
+            # Add rows to buffer
+            items = list(row_as_dict.values())
+            expected_rows = n_rows + len(items[0])
+            last_row = len(items[0])
+
+            if self.capacity <= expected_rows:
+                buffer_full = True
+                last_row = self.capacity-n_rows
+                expected_rows = self.capacity
+            if self.buffer is None:
+                # Initialize buffer as a list of empty tensors
+                self.buffer = []
+                for v in items:
+                    self.buffer.append(torch.empty((self.capacity,) + v.shape[1:], dtype=v.dtype, device=v.device))
+            # Copy new items into buffer
+            for k, v in enumerate(items):
+                self.buffer[k][n_rows:expected_rows] = v[:last_row]
+            n_rows = expected_rows
+
+        # At this point, dataloader has enough rows storted in memory.
+        # Set last_row_consumed = True in reader rather than draining the remainder
+        # of reader. If reader has infinite epochs, draining will be deadlock.
+        for row in self.reader:
+            self.reader.last_row_consumed = True
+        # Truncate empty tensors if capacity is larger than total rows.
+        if n_rows < self.capacity:
+            for k, v in enumerate(self.buffer):
+                self.buffer[k] = self.buffer[k][:n_rows]
+
+    def _iter_impl(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        self._load_rows_into_mem()
+
+        for epoch in range(self.num_epochs):
+            self.size = len(self.buffer[0])
```

Comment:
`size` can be a local variable. No reason for it to be the class state from what I see.